Algorithms/Algorithms/SplayTree/SplayTree.cs

```diff
@@ -0,0 +1,154 @@
+        /// <param name="searchValue">The value for which to search</param>
+        /// <returns>If no right Node exists, return the root. Otherwise, the root of the possibly rotated sub-tree</returns>
+        private static Node TraverseRightSubTree(Node root, T searchValue)
+        {
+            if (root.Right == null) return root;
+
+            if (root.Right.Value.CompareTo(searchValue) > 0)
+            {
+                root.Right.Left = Splay(root.Right.Left, searchValue);
+
+                if (root.Right.Left != null)
+                {
+                    root.Right = RightRotate(root.Right);
+                }
+            }
+            else if (root.Right.Value.CompareTo(searchValue) < 0)
+            {
+                root.Right.Right = Splay(root.Right.Right, searchValue);
+                root = LeftRotate(root);
+            }
+
+            return root.Right == null ? root : LeftRotate(root);
+        }
+
+        /// <summary>
+        /// Traverse through the left sub-tree of the given tree, looking for the search value.
+        /// </summary>
+        /// <param name="root">The root Node of the sub-tree to search</param>
+        /// <param name="searchValue">The value for which to search</param>
+        /// <returns>If no left Node exists, return the root. Otherwise, the root of the possibly rotated sub-tree</returns>
+        private static Node TraverseLeftSubTree(Node root, T searchValue)
+        {
+            if (root.Left == null) return root;
+
+            if (root.Left.Value.CompareTo(searchValue) > 0)
+            {
+                root.Left.Left = Splay(root.Left.Left, searchValue);
+                root = RightRotate(root);
+            }
+            else if (root.Left.Value.CompareTo(searchValue) < 0)
+            {
+                root.Left.Right = Splay(root.Left.Right, searchValue);
+
+                if (root.Left.Right != null)
+                {
+                    root.Left = LeftRotate(root.Left);
+                }
+            }
+
+            return root.Left == null ? root : RightRotate(root);
+        }
+
+        /// <summary>
+        /// Left rotate sub-tree rooted with x.
+        /// </summary>
+        /// <param name="x">Node around which to rotate</param>
+        /// <returns>The new root of the tree</returns>
+        private static Node LeftRotate(Node x)
+        {
+            var y = x.Right;
+            x.Right = y.Left;
+            y.Left = x;
+            return y;
+        }
+
+        /// <summary>
+        /// Right rotate sub-tree rooted with x.
+        /// </summary>
+        /// <param name="x">Node around which to rotate</param>
+        /// <returns>The new root of the tree</returns>
+        private static Node RightRotate(Node x)
+        {
+            var y = x.Left;
+            x.Left = y.Right;
+            y.Right = x;
+            return y;
+        }
+
+        #endregion
+
+        #region Classes
+
+        public class Node
```

Comment:
Public classes should be placed in separated files

Reply:
Done